### PR TITLE
add a plugin that opens spotify links in the client

### PR DESCRIPTION
**Pull request by crwn1337**
i miss this plugin from betterdiscord

### CODE REVIEW

**Comment by Vendicated** [comment]
this will only work for messages, not other links like bios. I think a click listener on the window that checks if the target is an a would be better as it covers all links

**Comment by crwn1337** [comment]
im guessing you mean something like this?

also, [SpotifyStore](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/SpotifyStore.ts#L91) would need to be updated, maybe [somehow use an href here](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/PlayerComponent.tsx#L288) instead of an onclick?

i tried changing it but i hate working with js/ts in general.

**Comment by Vendicated** [comment]
> also, [SpotifyStore](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/SpotifyStore.ts#L91) would need to be updated, maybe [somehow use an href here](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/PlayerComponent.tsx#L288) instead of an onclick?

image

**Comment by crwn1337** [comment]
> > also, [SpotifyStore](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/SpotifyStore.ts#L91) would need to be updated, maybe [somehow use an href here](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/PlayerComponent.tsx#L288) instead of an onclick?
> 
> image

i meant that option **should** become obsolete after this plugin, but the spotifystore plugin would need to be updated to use hyperlinks instead of `openExternal`

**Comment by Vendicated** [comment]
> > > also, [SpotifyStore](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/SpotifyStore.ts#L91) would need to be updated, maybe [somehow use an href here](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/PlayerComponent.tsx#L288) instead of an onclick?
> > 
> > 
> > image
> 
> i meant that option **should** become obsolete after this plugin, but the SpotifyControls plugin would need to be updated to use hyperlinks instead of `openExternal`

would you like to make this change?

File: src/plugins/openSpotifyLinks.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import definePlugin from "@utils/types";
+
+export default definePlugin({
+    name: "OpenSpotifyLinks",
+    description: "Opens Spotify links in the Spotify client.",
+    authors: [Devs.crwn],
+    patches: [],
+
+    start() {
+        this.clickListener = (event: MouseEvent) => {
```
Comment (Vendicated) [inline src/plugins/openSpotifyLinks.ts:29]: this assignment is obsolete, why not just make it a proper instance method? As in
```ts
onClick(event: MouseEvent) {
...
},

start() {
...
}
```

File: src/plugins/openSpotifyLinks.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+*/
+
+import { Devs } from "@utils/constants";
+import definePlugin from "@utils/types";
+
+export default definePlugin({
+    name: "OpenSpotifyLinks",
+    description: "Opens Spotify links in the Spotify client.",
+    authors: [Devs.crwn],
+    patches: [],
```
Comment (Vendicated) [inline src/plugins/openSpotifyLinks.ts:26]:
```suggestion
```

File: src/plugins/openSpotifyLinks.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    start() {
+        this.clickListener = (event: MouseEvent) => {
+            const target = event.target as HTMLElement;
+            if (!(target instanceof HTMLAnchorElement)) return;
+
+            const spotifyLinkRegex = /^https?:\/\/open\.spotify\.com\/(track|album|playlist|artist)\/([a-zA-Z0-9]+)(\?.*)?$/;
```
Comment (Vendicated) [inline src/plugins/openSpotifyLinks.ts:33]: please move this regex to the top level

File: src/plugins/openSpotifyLinks.ts (hunk 1 above)
Comment (crwn1337) [reply src/plugins/openSpotifyLinks.ts]: I'm guessing you mean like this? Since onClick isn't a method that's called by Vencord

**Comment by crwn1337** [comment]
> > > > also, [SpotifyStore](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/SpotifyStore.ts#L91) would need to be updated, maybe [somehow use an href here](https://github.com/Vendicated/Vencord/blob/0b2c3c834a74fb84dc8cb6b22589530f3420828b/src/plugins/spotifyControls/PlayerComponent.tsx#L288) instead of an onclick?
> > > 
> > > 
> > > image
> > 
> > 
> > i meant that option **should** become obsolete after this plugin, but the SpotifyControls plugin would need to be updated to use hyperlinks instead of `openExternal`
> 
> would you like to make this change?

I tried, I wasn't able to do it.

**Comment by crwn1337** [comment]
https://github.com/Vendicated/Vencord/commit/088a8bd1b681b6983c19a8b062e7f755374a4335

**Comment by Vendicated** [comment]
oh I'm really sorry, I entirely forgot this pr was a thing. Thanks anyway for your effort

**Pull request closed without merging**